File: crates/overlay/src/projection.rs
//! TASK 3 — Pinhole-camera projection: ETS2 world-space → screen pixels.
//!
//! ## Coordinate conventions
//!
//! **ETS2 world space** (left-handed, as described in coords.rs):
//!   X = East (+), Y = Up (+), Z = South (+) / North (−)
//!
//! **Camera space** (right-handed, Z = depth / forward into screen):
//!   The heading quaternion from `ets2_heading_to_quat` maps ETS2 world
//!   vectors into a camera frame where +Z is the look-at direction.
//!
//! ## Heading note
//! The spec states "+Z forward at heading=0". TruckPilot's own docs state
//! "heading=0 = North = −Z". If live testing shows AR lines are 180° off,
//! flip heading with `(heading + 0.5) % 1.0` in `ar_renderer.rs`.
//! The FOV calibration wizard (Task 7) cannot fix a heading flip.

use glam::{EulerRot, Quat, Vec3};

// ─── camera intrinsics ───────────────────────────────────────────────────────

/// Pinhole camera intrinsic matrix (simplified: fx == fy, principal point centered).
#[derive(Debug, Clone, Copy)]
pub struct CameraIntrinsics {
    /// Focal length in pixels (same for X and Y — square pixels assumed).
    pub f_x: f32,
    /// Principal point X (screen centre, pixels).
    pub cx: f32,
    /// Principal point Y (screen centre, pixels).
    pub cy: f32,
}

impl CameraIntrinsics {
    /// Build from horizontal FOV angle and screen dimensions.
    pub fn from_hfov(fov_h_deg: f32, w: f32, h: f32) -> Self {
        let f = (w / 2.0) / (fov_h_deg.to_radians() / 2.0).tan();
        Self { f_x: f, cx: w / 2.0, cy: h / 2.0 }
    }
}

// ─── rotation helper ─────────────────────────────────────────────────────────

/// Convert SCS SDK euler orientation to a glam quaternion.
///
/// All three values are in the 0..1 range (SCS SDK unit where 1.0 = full turn).
///
/// The returned quaternion satisfies: `q * [0,0,1]` = look-at direction in world space.
///
/// ETS2 heading convention (from lane-follower: `(-heading * 360).rem_euclid(360)` = CW°):
///   heading=0   → North (−Z)   heading=0.25 → West (−X)
///   heading=0.5 → South (+Z)   heading=0.75 → East (+X)
///
/// Required Y-rotation for look-at [−sin(h), 0, −cos(h)] = R_y(π + h) * [0,0,1].
/// Derivation: sin(π+h)=−sin(h) ✓  cos(π+h)=−cos(h) ✓
///
/// Note: the original spec used `−h` which assumed heading=0 = South (+Z).
/// Corrected to `π + h` to match TruckPilot's heading=0 = North = −Z convention.
pub fn ets2_heading_to_quat(heading: f32, pitch: f32, roll: f32) -> Quat {
    use std::f32::consts::PI;
    let h = heading * 2.0 * PI;
    let p = pitch * 2.0 * PI;
    let r = roll * 2.0 * PI;
    // YXZ: apply yaw (Y) first. (π + h) maps heading=0/North to look-at −Z.
    Quat::from_euler(EulerRot::YXZ, PI + h, p, r)
}

// ─── projection ──────────────────────────────────────────────────────────────

/// Project a single world-space point onto the screen.
///
/// Returns `None` when the point is at or behind the near plane.
pub fn world_to_screen(
    world_pt: Vec3,
    cam_pos: Vec3,
    cam_rot: Quat,
    k: &CameraIntrinsics,
    near: f32,
) -> Option<(f32, f32)> {
    let p_cam = cam_rot.inverse().mul_vec3(world_pt - cam_pos);
    if p_cam.z < near {
        return None;
    }
    let px = k.f_x * (p_cam.x / p_cam.z) + k.cx;
    let py = k.cy - k.f_x * (p_cam.y / p_cam.z); // screen Y inverted vs camera Y
    Some((px, py))
}

// ─── near-plane clipping ─────────────────────────────────────────────────────

/// Clip a line segment so neither endpoint is behind the near plane.
///
/// Both input points are **already in camera space** (result of `cam_rot.inverse() * delta`).
/// Returns `None` when the whole segment is behind `near`.
/// When one endpoint is behind, linearly interpolates to the near plane.
pub fn clip_segment_to_near(
    p0: Vec3,
    p1: Vec3,
    near: f32,
) -> Option<(Vec3, Vec3)> {
    let p0_behind = p0.z < near;
    let p1_behind = p1.z < near;

    match (p0_behind, p1_behind) {
        (true, true) => None,
        (false, false) => Some((p0, p1)),
        (true, false) => {
            // p0 behind → clip p0 to near plane
            let t = (near - p0.z) / (p1.z - p0.z);
            Some((p0.lerp(p1, t), p1))
        }
        (false, true) => {
            // p1 behind → clip p1 to near plane
            let t = (near - p0.z) / (p1.z - p0.z);
            Some((p0, p0.lerp(p1, t)))
        }
    }
}

// ─── tests ───────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    const W: f32 = 1920.0;
    const H: f32 = 1080.0;
    const FOV: f32 = 75.0;
    const NEAR: f32 = 0.1;

    fn k() -> CameraIntrinsics {
        CameraIntrinsics::from_hfov(FOV, W, H)
    }

    /// Identity cam_rot (looking along +Z) — used for spec-matching tests.
    fn identity_cam() -> (Vec3, Quat) {
        (Vec3::ZERO, Quat::IDENTITY)
    }

    #[test]
    fn test_world_pt_directly_ahead_projects_to_center() {
        // When cam_rot = identity, "ahead" = +Z direction.
        let (cam_pos, cam_rot) = identity_cam();
        let pt = Vec3::new(0.0, 0.0, 10.0); // 10 m along +Z
        let (px, py) = world_to_screen(pt, cam_pos, cam_rot, &k(), NEAR).unwrap();
        assert!((px - W / 2.0).abs() < 1.0, "should project to cx, got {px}");
        assert!((py - H / 2.0).abs() < 1.0, "should project to cy, got {py}");
    }

    #[test]
    fn test_world_pt_above_camera_projects_above_cy() {
        let (cam_pos, cam_rot) = identity_cam();
        // Point above camera centre (+Y) at depth 10 m → pixel Y < cy (above centre)
        let pt = Vec3::new(0.0, 1.0, 10.0);
        let (_, py) = world_to_screen(pt, cam_pos, cam_rot, &k(), NEAR).unwrap();
        assert!(py < H / 2.0, "point above cam should project above cy, got py={py}");
    }

    #[test]
    fn test_world_pt_behind_camera_returns_none() {
        let (cam_pos, cam_rot) = identity_cam();
        let pt = Vec3::new(0.0, 0.0, -1.0); // behind camera (z < near)
        assert!(
            world_to_screen(pt, cam_pos, cam_rot, &k(), NEAR).is_none(),
            "point behind camera should return None"
        );
    }

    #[test]
    fn test_clip_segment_one_endpoint_behind() {
        // p0 is behind near (z=0 < 0.1), p1 is in front (z=10)
        let p0 = Vec3::new(0.0, 0.0, 0.0);
        let p1 = Vec3::new(1.0, 0.0, 10.0);
        let (a, b) = clip_segment_to_near(p0, p1, NEAR).unwrap();
        // Clipped a.z should equal near
        assert!((a.z - NEAR).abs() < 1e-5, "clipped z should be near={NEAR}, got {}", a.z);
        // b unchanged
        assert!((b - p1).length() < 1e-5, "far endpoint unchanged");
    }

    #[test]
    fn test_clip_segment_both_behind_returns_none() {
        let p0 = Vec3::new(0.0, 0.0, -5.0);
        let p1 = Vec3::new(1.0, 0.0, -1.0);
        assert!(clip_segment_to_near(p0, p1, NEAR).is_none());
    }

    #[test]
    fn test_ets2_heading_to_quat_at_0_returns_defined_rotation() {
        // Heading=0 (North = −Z): quaternion must be normalised
        let q = ets2_heading_to_quat(0.0, 0.0, 0.0);
        assert!(
            (q.length() - 1.0).abs() < 1e-5,
            "quaternion must be unit, length={}", q.length()
        );
    }

    #[test]
    fn test_heading_0_looks_north() {
        // heading=0 = North = (0, 0, −1) in ETS2 world space.
        // cam_rot * [0,0,1] should give the look-at direction = [0,0,−1].
        let q = ets2_heading_to_quat(0.0, 0.0, 0.0);
        let look_at = q.mul_vec3(glam::Vec3::Z);
        assert!(look_at.z < -0.99, "heading=0 should look North (−Z), got {look_at:?}");
        assert!(look_at.x.abs() < 0.01, "heading=0 should have no X component, got {look_at:?}");
    }

    #[test]
    fn test_heading_75_looks_east() {
        // heading=0.75 = East = (+1, 0, 0).
        let q = ets2_heading_to_quat(0.75, 0.0, 0.0);
        let look_at = q.mul_vec3(glam::Vec3::Z);
        assert!(look_at.x > 0.99, "heading=0.75 should look East (+X), got {look_at:?}");
        assert!(look_at.z.abs() < 0.01, "heading=0.75 should have no Z component");
    }

    #[test]
    fn intrinsics_principal_point_at_center() {
        let ki = CameraIntrinsics::from_hfov(90.0, 1920.0, 1080.0);
        assert!((ki.cx - 960.0).abs() < 1e-3);
        assert!((ki.cy - 540.0).abs() < 1e-3);
    }

    #[test]
    fn intrinsics_fov90_focal_length() {
        // At FOV=90°: f = (w/2) / tan(45°) = (w/2) / 1.0 = w/2
        let ki = CameraIntrinsics::from_hfov(90.0, 1920.0, 1080.0);
        assert!((ki.f_x - 960.0).abs() < 1.0, "f_x={}", ki.f_x);
    }
}

File: crates/ui/src/lib/projection.ts
// Phase 6.6a-1 — Clean-room world→screen projection for the Tauri overlay.
//
// CONVENTION — re-derived here from standard pinhole geometry. The axis/angle
// convention is ADOPTED from crates/overlay/src/projection.rs (read as a
// reference only; no code copied) so this TS projection matches the existing
// DX11 renderer and the line does not end up mirrored/rotated:
//
//   ETS2 world space (left-handed): X = East(+), Y = Up(+), Z = South(+)/North(−).
//   heading/pitch/roll are SCS-SDK units in 0..1 turns (× 2π → radians).
//   Camera rotation R (camera→world) = Ry(π + heading)·Rx(pitch)·Rz(roll)
//     (glam EulerRot::YXZ, intrinsic). The +π yaw maps heading=0 ⇒ look −Z
//     (North) and heading=0.75 ⇒ look +X (East).
//   world→camera = Rᵀ. Pinhole: +Z is depth into the screen, screen-Y inverted.
//
// SCREEN-X IS NEGATED vs the reference (px = cx − f·x/z, not +). ETS2 world is
// left-handed (X=East), so the right-handed camera math mirrors the lateral
// axis: a point to the physical right (East when facing North) would land LEFT
// of centre. The reference (crates/overlay/src/projection.rs + ar_renderer.rs)
// leaves this mirror in — its own diag.rs MIRROR-CHECK documents it ("px < cx ⇒
// CAMERA-X MIRROR"), but the renderer was never fixed because a symmetric road
// ribbon hides it. A discrete marker exposes it (wrong side, slides off-road
// when steering), so we un-mirror here. This is a DELIBERATE divergence from
// the reference's world_to_screen; the rotation/inversion itself is unchanged.
//
// NOTE on pitch/roll units: the reference treats ALL three euler values as
// 0..1 turns (×2π). On flat ground pitch≈roll≈0, so the 6.6a-1 test marker is
// unaffected; verify the unit on a real slope before trusting tilt accuracy.

export type Vec3 = readonly [number, number, number];

/** Fixed horizontal field of view (deg). ETS2's real cockpit FOV differs and
 *  varies per truck/zoom — this is a first guess; the reference renderer ships
 *  a calibration wizard for the same reason. Promote to a slider later. */
export const FOV_H_DEG = 75;
/** Near plane (m). Points at/behind this are not drawn. Matches the reference. */
export const NEAR = 0.1;
/** Cabin camera offset above the truck reference point (m).
 *  From crates/overlay/src/telemetry.rs (CABIN_Y_OFFSET). */
export const CABIN_Y_OFFSET = 1.5;
/** Cabin camera offset forward along heading (m).
 *  From crates/overlay/src/telemetry.rs (CABIN_FORWARD_OFFSET). */
export const CABIN_FORWARD_OFFSET = 0.5;

const TAU = Math.PI * 2;

export interface CameraIntrinsics {
  /** Focal length in pixels (fx == fy, square pixels assumed). */
  readonly f: number;
  /** Principal point X (screen centre), pixels. */
  readonly cx: number;
  /** Principal point Y (screen centre), pixels. */
  readonly cy: number;
}

/** Build intrinsics from horizontal FOV and the canvas size in CSS pixels. */
export function cameraIntrinsics(fovHDeg: number, w: number, h: number): CameraIntrinsics {
  const f = w / 2 / Math.tan((fovHDeg * (Math.PI / 180)) / 2);
  return { f, cx: w / 2, cy: h / 2 };
}

/** Unit forward direction on the ground plane for a heading (SCS 0..1 turns).
 *  forward = (−sin h, 0, −cos h): heading=0 → (0,0,−1)=North, 0.75 → (1,0,0)=East. */
export function headingForward(heading: number): Vec3 {
  const h = heading * TAU;
  return [-Math.sin(h), 0, -Math.cos(h)];
}

/** Cabin/head camera position from the truck reference point + heading.
 *  Mirrors compute_head_pos() in crates/overlay/src/telemetry.rs. */
export function cabinCameraPos(truck: Vec3, heading: number): Vec3 {
  const h = heading * TAU;
  return [
    truck[0] - Math.sin(h) * CABIN_FORWARD_OFFSET,
    truck[1] + CABIN_Y_OFFSET,
    truck[2] - Math.cos(h) * CABIN_FORWARD_OFFSET,
  ];
}

/** A world point `dist` metres ahead of the truck along its heading, at the
 *  truck's ground height (truck Y). Used to drop the 6.6a-1 test markers. */
export function pointAhead(truck: Vec3, heading: number, dist: number): Vec3 {
  const fwd = headingForward(heading);
  return [truck[0] + fwd[0] * dist, truck[1], truck[2] + fwd[2] * dist];
}

// ── rotation ──────────────────────────────────────────────────────────────
// We only ever need Rᵀ·d (world→camera), so we build R = Ry(yaw)·Rx(p)·Rz(r)
// from elementary right-handed matrices (glam convention) and apply Rᵀ
// directly — no matrix inversion needed (R is orthonormal, so R⁻¹ = Rᵀ).

/** Transform a world-space delta (worldPt − camPos) into camera space using the
 *  camera's heading/pitch/roll (SCS 0..1 turns). Returns [x,y,z] with +z = depth. */
function worldDeltaToCamera(d: Vec3, heading: number, pitch: number, roll: number): Vec3 {
  const yaw = Math.PI + heading * TAU;
  const p = pitch * TAU;
  const r = roll * TAU;

  const cy = Math.cos(yaw);
  const sy = Math.sin(yaw);
  const cp = Math.cos(p);
  const sp = Math.sin(p);
  const cr = Math.cos(r);
  const sr = Math.sin(r);

  // Rx·Rz, row-major (m<row><col>):
  //   Rx = [[1,0,0],[0,cp,-sp],[0,sp,cp]]
  //   Rz = [[cr,-sr,0],[sr,cr,0],[0,0,1]]
  const m00 = cr;
  const m01 = -sr;
  const m02 = 0;
  const m10 = cp * sr;
  const m11 = cp * cr;
  const m12 = -sp;
  const m20 = sp * sr;
  const m21 = sp * cr;
  const m22 = cp;

  // R = Ry·(Rx·Rz), with Ry = [[cy,0,sy],[0,1,0],[-sy,0,cy]] (r<row><col>).
  const r00 = cy * m00 + sy * m20;
  const r01 = cy * m01 + sy * m21;
  const r02 = cy * m02 + sy * m22;
  const r10 = m10;
  const r11 = m11;
  const r12 = m12;
  const r20 = -sy * m00 + cy * m20;
  const r21 = -sy * m01 + cy * m21;
  const r22 = -sy * m02 + cy * m22;

  // p_cam = Rᵀ·d  →  p_cam[i] = Σ_j R[j][i]·d[j].
  const [d0, d1, d2] = d;
  return [
    r00 * d0 + r10 * d1 + r20 * d2,
    r01 * d0 + r11 * d1 + r21 * d2,
    r02 * d0 + r12 * d1 + r22 * d2,
  ];
}

/** Transform a world-space point into camera space (before projection). +z is
 *  depth into the screen. Exposed so callers can near-clip a line segment in
 *  camera space (see clipSegmentToNear) before projecting its endpoints. */
export function worldToCamera(
  worldPt: Vec3,
  camPos: Vec3,
  heading: number,
  pitch: number,
  roll: number,
): Vec3 {
  const d: Vec3 = [worldPt[0] - camPos[0], worldPt[1] - camPos[1], worldPt[2] - camPos[2]];
  return worldDeltaToCamera(d, heading, pitch, roll);
}

export interface ScreenPoint {
  x: number;
  y: number;
  /** Camera-space Z (metres ahead of the camera). */
  depth: number;
}

/** Project an already-in-camera-space point to screen pixels. Returns null at
 *  or behind the near plane. Screen-X is negated (left-handed world un-mirror,
 *  see module header). */
export function projectCameraPoint(
  c: Vec3,
  intr: CameraIntrinsics,
  near: number = NEAR,
): ScreenPoint | null {
  if (c[2] < near) return null;
  return {
    x: intr.cx - intr.f * (c[0] / c[2]),
    y: intr.cy - intr.f * (c[1] / c[2]),
    depth: c[2],
  };
}

/** Project a world-space point onto the screen. Returns null when the point is
 *  at or behind the near plane. `depth` is the camera-space Z (metres ahead). */
export function worldToScreen(
  worldPt: Vec3,
  camPos: Vec3,
  heading: number,
  pitch: number,
  roll: number,
  intr: CameraIntrinsics,
  near: number = NEAR,
): ScreenPoint | null {
  return projectCameraPoint(worldToCamera(worldPt, camPos, heading, pitch, roll), intr, near);
}

/** Clip a line segment whose endpoints are ALREADY in camera space to the near
 *  plane. Returns null when both endpoints are behind `near`; otherwise the
 *  segment with any behind-endpoint moved onto the near plane (linear interp).
 *  Mirrors clip_segment_to_near in crates/overlay/src/projection.rs. Clipping
 *  in camera space (not screen space) is what stops a behind-camera endpoint
 *  from projecting to a garbage pixel and streaking across the frame. */
export function clipSegmentToNear(
  c0: Vec3,
  c1: Vec3,
  near: number = NEAR,
): [Vec3, Vec3] | null {
  const behind0 = c0[2] < near;
  const behind1 = c1[2] < near;
  if (behind0 && behind1) return null;
  if (!behind0 && !behind1) return [c0, c1];
  // Exactly one endpoint behind → move it onto the near plane along the segment.
  const t = (near - c0[2]) / (c1[2] - c0[2]);
  const mid: Vec3 = [
    c0[0] + (c1[0] - c0[0]) * t,
    c0[1] + (c1[1] - c0[1]) * t,
    c0[2] + (c1[2] - c0[2]) * t,
  ];
  return behind0 ? [mid, c1] : [c0, mid];
}
